onnxruntime/test/python/transformers/test_gqa_cpu.py

```diff
@@ -0,0 +1,1884 @@
+    num_heads = 0
+    kv_num_heads = 0
+    head_size = 0
+
+    def __init__(self, b, sq, skv, sb, n, n2, h):
+        self.batch_size = b
+        self.q_sequence_length = sq
+        self.kv_sequence_length = skv
+        self.buffer_sequence_length = sb
+        self.num_heads = n
+        self.kv_num_heads = n2
+        self.head_size = h
+
+
+# LLaMA Microsoft model
+class LlamaMSRotaryEmbedding(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+
+    def rotate_tensor(
+        self,
+        x: torch.Tensor,  # BxSxNxH
+        cos: torch.Tensor,  # 1xSx1x(H/2)
+        sin: torch.Tensor,  # 1xSx1x(H/2)
+        pos: torch.Tensor,
+        interleaved: bool,
+    ):
+        # Dimension of x is [batch_size, seq_len, n_heads, head_dim]
+        rot_dim = 2 * cos.shape[3]
+
+        # Dolly requires partial rotation
+        x_rot = x[:, :, :, :rot_dim]
+
+        if interleaved:
+            x1 = x_rot[:, :, :, 0::2]
+            x2 = x_rot[:, :, :, 1::2]
+        else:
+            half = x_rot.shape[-1] // 2
+            x1 = x[:, :, :, 0:half]
+            x2 = x[:, :, :, half : 2 * half]
+
+        seq_len = x.shape[1]
+
+        # cos_x: (1, S, 1, H/2)
+        # sin_x: (1, S, 1, H/2)
+        # x1: (B, S, N, H/2)
+        # x2: (B, S, N, H/2)
+        if seq_len == 1:
+            batch_size = x.shape[0]
+            pos_i = pos.unsqueeze(1).unsqueeze(2).unsqueeze(3).long()
+            cos_x = cos.expand(batch_size, -1, -1, -1)
+            sin_x = sin.expand(batch_size, -1, -1, -1)
+            cos_x = cos_x.gather(1, pos_i.expand(-1, -1, cos.shape[2], cos.shape[3]))
+            sin_x = sin_x.gather(1, pos_i.expand(-1, -1, sin.shape[2], sin.shape[3]))
+            real = cos_x * x1 - sin_x * x2
+            imag = sin_x * x1 + cos_x * x2
+            if interleaved:
+                x_rot[:, :, :, 0::2] = real
+                x_rot[:, :, :, 1::2] = imag
+            else:
+                x_rot = torch.cat((real, imag), dim=-1)
+        else:
+            cos_x = cos[:, 0:seq_len, :, :]
+            sin_x = sin[:, 0:seq_len, :, :]
+            real = cos_x * x1 - sin_x * x2
+            imag = sin_x * x1 + cos_x * x2
+            if interleaved:
+                x_rot[:, :, :, 0::2] = real
+                x_rot[:, :, :, 1::2] = imag
+            else:
+                x_rot = torch.cat((real, imag), dim=-1)
+
+        return torch.cat((x_rot, x[:, :, :, rot_dim:]), dim=-1)
+
+    def forward(self, x, cos, sin, pos, interleaved):
+        return self.rotate_tensor(x, cos, sin, pos, interleaved)
+
+
+def create_group_query_attention_graph_prompt(
+    config,
+    past_kv_format=Formats.BSNH,
+    share_buffer=True,
+    local_window_size=-1,
+    rotary=False,
+    rotary_interleaved=False,
+    packed=False,
+):
+    past_kv_seqlen = config.buffer_sequence_length if share_buffer else 0
+    present_kv_seqlen = config.buffer_sequence_length if share_buffer else config.kv_sequence_length
+    nodes = [
```

Comment:
Hi, @aciddelgado, I didn't understand why in prompt mode, we need test past_key and past_value. Does this overlap with past mode?

Reply:
Thank you!